homeassistant/components/synology_dsm/media_source.py

```diff
@@ -0,0 +1,232 @@
+            if mime_type.startswith("image/"):
+                # Force small small thumbnails
+                album_item.thumbnail_size = "sm"
+                ret.append(
+                    BrowseMediaSource(
+                        domain=DOMAIN,
+                        identifier=f"{identifier.unique_id}/{identifier.album_id}/{album_item.thumbnail_cache_key}/{album_item.file_name}",
+                        media_class=MediaClass.IMAGE,
+                        media_content_type=mime_type,
+                        title=album_item.file_name,
+                        can_play=True,
+                        can_expand=False,
+                        thumbnail=await self.async_get_thumbnail(
+                            album_item, diskstation
+                        ),
+                    )
+                )
+        return ret
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> PlayMedia:
+        """Resolve media to a url."""
+        identifier = SynologyPhotosMediaSourceIdentifier(item.identifier)
+        if identifier.album_id is None:
+            raise Unresolvable("No album id")
+        if identifier.file_name is None:
+            raise Unresolvable("No file name")
+        mime_type, _ = mimetypes.guess_type(identifier.file_name)
+        if not isinstance(mime_type, str):
+            raise Unresolvable("No file extension")
+        return PlayMedia(
+            f"/synology_dsm/{identifier.unique_id}/{identifier.cache_key}/{identifier.file_name}",
+            mime_type,
+        )
+
+    async def async_get_thumbnail(
+        self, item: SynoPhotosItem, diskstation: SynologyDSMData
+    ) -> str:
+        """Get thumbnail."""
+        try:
+            thumbnail = await diskstation.api.photos.get_item_thumbnail_url(item)
+        except SynologyDSMException:
+            return ""
+        return str(thumbnail)
+
+
+class SynologyDsmMediaView(http.HomeAssistantView):
+    """Synology Media Finder View."""
+
+    url = "/synology_dsm/{source_dir_id}/{location:.*}"
+    name = "synology_dsm"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the media view."""
+        self.hass = hass
+
+    async def get(
+        self, request: web.Request, source_dir_id: str, location: str
+    ) -> web.Response:
+        """Start a GET request."""
+        if not self.hass.data.get(DOMAIN):
+            raise web.HTTPNotFound()
+        # location: {cache_key}/{filename}
+        cache_key, file_name = location.split("/")
+        image_id = cache_key.split("_")[0]
+        file_extension = file_name.split(".")[-1]
+        mime_type, _ = mimetypes.guess_type(file_name)
+        if not isinstance(mime_type, str):
+            raise web.HTTPNotFound()
+        diskstation: SynologyDSMData = self.hass.data[DOMAIN][source_dir_id]
+
+        item = SynoPhotosItem(image_id, "", "", "", cache_key, "")
+        try:
+            image = await diskstation.api.photos.download_item(item)
```

Comment:
Is it possible that this API can access items other than photos. Is there any risk of a directory transversal attack here?

Reply:
The synology photos application also does server side checks. You can't access pictures that aren't yours.